group_theory/perm.lean

```diff
@@ -286,9 +385,204 @@ funext $ λ f,
+    exact this _),
+let ⟨z, hz⟩ := hf in
+let ⟨i, hi⟩ := hz.2 x hfx in
+if hyx : y = x then by simp [hyx]
+else if hfyx : y = f x then by simp [hfyx, hffx]
+else begin
+  rw [swap_apply_of_ne_of_ne hyx hfyx],
+  refine by_contradiction (λ hy, _),
+  cases hz.2 y hy with j hj,
+  rw [← sub_add_cancel j i, gpow_add, mul_apply, hi] at hj,
+  cases this (j - i) with hji hji,
+  { rw [← hj, hji] at hyx, cc },
+  { rw [← hj, hji] at hfyx, cc }
+end
+
+lemma is_cycle_swap_mul {f : perm α} (hf : is_cycle f) {x : α}
+  (hx : f x ≠ x ∧ ∀ y, f y ≠ y → ∃ i : ℤ, (f ^ i) x = y) :
+  is_cycle (swap x (f x) * f) ∨ swap x (f x) * f = 1 :=
+if hffx : f (f x) = x
+then or.inr $ calc
+  swap x (f x) * f = swap x (f x) * swap x (f x) :
+    by rw ← eq_swap_of_is_cycle_of_apply_apply_eq_self hf hx.1 hffx
+  ... = 1 : swap_swap _ _
+else or.inl
+  ⟨f x, by simp only [swap_apply_def, mul_apply];
+          split_ifs; simp [injective.eq_iff f.bijective.1] at *; cc,
+    λ y hy,
+    let ⟨i, hi⟩ := hx.2 _ (support_swap_mul hy).1 in
+    have hi : (f ^ (i - 1)) (f x) = y, from
+      calc (f ^ (i - 1)) (f x) = (f ^ (i - 1) * f ^ (1 : ℤ)) x : by rw [gpow_one, mul_apply]
+      ... =  y : by rwa [← gpow_add, sub_add_cancel],
+    is_cycle_swap_mul_aux (i - 1) hf hy hi⟩
+
+lemma support_swap_mul_cycle [fintype α] {f : perm α} (hf : is_cycle f) {x : α}
+  (hx : f x ≠ x ∧ ∀ y, f y ≠ y → ∃ i : ℤ, (f ^ i) x = y) :
+  (swap x (f x) * f).support = f.support.erase x ∨ swap x (f x) * f = 1 :=
+or_iff_not_imp_right.2 (λ h1, finset.ext.2 $ λ y, begin
+  have hfyxor : f y ≠ x ∨ f x ≠ y :=
+    not_and_distrib.1 (λ h, h1 $
+      by have := eq_swap_of_is_cycle_of_apply_apply_eq_self hf hx.1 (by rw [h.2, h.1]);
+      rw [← this, this, mul_def, swap_swap, one_def]),
+  rw [mem_support, mem_erase, mem_support],
+  split,
+  { assume h,
+    refine not_or_distrib.1 (λ h₁, h₁.elim
+      (λ hyx, by simpa [hyx, mul_apply] using h) _),
+    assume hfy,
+    have hyx : x ≠ y := λ h, by rw h at hx; tauto,
+    have hfyx : f x ≠ y := by rwa [← hfy, ne.def, injective.eq_iff f.bijective.1],
+    simpa [mul_apply, hfy, swap_apply_of_ne_of_ne hyx.symm hfyx.symm] using h },
+  { assume h,
+    simp [swap_apply_def],
+    split_ifs; cc }
+end)
+
+@[simp] lemma support_swap [fintype α] {x y : α} (hxy : x ≠ y) : (swap x y).support = {x, y} :=
+finset.ext.2 $ λ a, by simp [swap_apply_def]; split_ifs; cc
+
+lemma card_support_swap [fintype α] {x y : α} (hxy : x ≠ y) : (swap x y).support.card = 2 :=
+show (swap x y).support.card = finset.card ⟨x::y::0, by simp [hxy]⟩,
+from congr_arg card $ by rw [support_swap hxy]; simp [*, finset.ext]; cc
+
+lemma sign_cycle [fintype α] : ∀ {f : perm α} (hf : is_cycle f),
+  sign f = -1 ^ (f.support.card + 1)
+| f := λ hf,
+let ⟨x, hx⟩ := hf in
+calc sign f = sign (swap x (f x) * (swap x (f x) * f)) :
+by rw [← mul_assoc, mul_def, mul_def, swap_swap, trans_refl]
+... = -1 ^ (f.support.card + 1) :
+if h1 : (swap x (f x) * f) = 1 then
+have h : f = swap x (f x),
```

Comment:
indent the `if` and then indent the part between `then` and `else`.

Reply:
I'm fine when you have a one-line proof and put `then` on the one-liner. But in this case you have a multi-line proof and it should be indented.